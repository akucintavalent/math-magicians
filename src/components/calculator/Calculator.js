import React from 'react';
import './calculator.css';
import calculate from '../../logic/calculate';

class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = { calc: {} };
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(e) {
    const { calc } = this.state;
    const buttonName = e.target.name;
    const newCalc = calculate(calc, buttonName);
    this.setState({ calc: newCalc });
  }

  render() {
    const {
      calc: {
        total,
        next,
        operation,
      },
    } = this.state;
    const display = ((total || '') + (operation || '') + (next || '')) || '0';
    return (
      <div id="calculator">
        <p id="input">{display}</p>
        <button type="button" onClick={this.handleClick} name="AC" id="ac">AC</button>
        <button type="button" onClick={this.handleClick} name="+/-" id="plus-minus">+/-</button>
        <button type="button" onClick={this.handleClick} name="%" id="percent">%</button>
        <button type="button" onClick={this.handleClick} name="รท" className="Operations" id="div">รท</button>
        <button type="button" onClick={this.handleClick} name="7" id="btn-7">7</button>
        <button type="button" onClick={this.handleClick} name="8" id="btn-8">8</button>
        <button type="button" onClick={this.handleClick} name="9" id="btn-9">9</button>
        <button type="button" onClick={this.handleClick} name="x" className="Operations" id="mult">&times;</button>
        <button type="button" onClick={this.handleClick} name="4" id="btn-4">4</button>
        <button type="button" onClick={this.handleClick} name="5" id="btn-5">5</button>
        <button type="button" onClick={this.handleClick} name="6" id="btn-6">6</button>
        <button type="button" onClick={this.handleClick} name="-" className="Operations" id="minus">-</button>
        <button type="button" onClick={this.handleClick} name="1" id="btn-1">1</button>
        <button type="button" onClick={this.handleClick} name="2" id="btn-2">2</button>
        <button type="button" onClick={this.handleClick} name="3" id="btn-3">3</button>
        <button type="button" onClick={this.handleClick} name="+" className="Operations" id="plus">+</button>
        <button type="button" onClick={this.handleClick} name="0" id="btn-0">0</button>
        <button type="button" onClick={this.handleClick} name="." id="dot">.</button>
        <button type="button" onClick={this.handleClick} name="=" className="Operations" id="equal">=</button>
      </div>
    );
  }
}

export default Calculator;
