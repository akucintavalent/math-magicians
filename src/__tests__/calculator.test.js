import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import Calculator from '../components/calculator/Calculator';

describe('Calculator component', () => {
  test('snapshot', () => {
    const tree = renderer
      .create(<Calculator />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('addition', async () => {
    const dom = render(<Calculator />);

    await userEvent.click(screen.getByText('4'));
    await userEvent.click(screen.getByText('6'));
    await userEvent.click(screen.getByText('2'));
    await userEvent.click(screen.getByText('+'));
    await userEvent.click(screen.getByText('9'));
    await userEvent.click(screen.getByText('0'));
    await userEvent.click(screen.getByText('='));
    const input = dom.container.querySelector('#input');
    expect(input.innerHTML).toBe('552');
  });

  test('multiplication', async () => {
    const dom = render(<Calculator />);

    await userEvent.click(screen.getByText('4'));
    await userEvent.click(screen.getByText('6'));
    await userEvent.click(screen.getByText('2'));
    await userEvent.click(screen.getByText('x'));
    await userEvent.click(screen.getByText('9'));
    await userEvent.click(screen.getByText('0'));
    await userEvent.click(screen.getByText('='));
    const input = dom.container.querySelector('#input');
    expect(input.innerHTML).toBe((462 * 90).toString());
  });

  test('not finished operation of multiplication', async () => {
    const dom = render(<Calculator />);

    await userEvent.click(screen.getByText('4'));
    await userEvent.click(screen.getByText('.'));
    await userEvent.click(screen.getByText('1'));
    await userEvent.click(screen.getByText('x'));
    await userEvent.click(screen.getByText('9'));
    await userEvent.click(screen.getByText('0'));
    const input = dom.container.querySelector('#input');
    expect(input.innerHTML).toBe('4.1x90');
  });

  test('division', async () => {
    const dom = render(<Calculator />);

    await userEvent.click(screen.getByText('1'));
    await userEvent.click(screen.getByText('8'));
    await userEvent.click(screen.getByText('0'));
    await userEvent.click(screen.getByText('รท'));
    await userEvent.click(screen.getByText('9'));
    await userEvent.click(screen.getByText('0'));
    await userEvent.click(screen.getByText('='));
    const input = dom.container.querySelector('#input');
    expect(input.innerHTML).toBe('2');
  });
});
