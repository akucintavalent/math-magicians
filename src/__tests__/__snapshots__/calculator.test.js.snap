// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component snapshot 1`] = `
<div
  id="calculator"
>
  <p
    id="input"
  >
    0
  </p>
  <button
    className=""
    name="AC"
    onClick={[Function]}
    type="button"
  >
    AC
  </button>
  <button
    className=""
    name="+/-"
    onClick={[Function]}
    type="button"
  >
    +/-
  </button>
  <button
    className=""
    name="%"
    onClick={[Function]}
    type="button"
  >
    %
  </button>
  <button
    className="Operations"
    name="รท"
    onClick={[Function]}
    type="button"
  >
    รท
  </button>
  <button
    className=""
    name="7"
    onClick={[Function]}
    type="button"
  >
    7
  </button>
  <button
    className=""
    name="8"
    onClick={[Function]}
    type="button"
  >
    8
  </button>
  <button
    className=""
    name="9"
    onClick={[Function]}
    type="button"
  >
    9
  </button>
  <button
    className="Operations"
    name="x"
    onClick={[Function]}
    type="button"
  >
    x
  </button>
  <button
    className=""
    name="4"
    onClick={[Function]}
    type="button"
  >
    4
  </button>
  <button
    className=""
    name="5"
    onClick={[Function]}
    type="button"
  >
    5
  </button>
  <button
    className=""
    name="6"
    onClick={[Function]}
    type="button"
  >
    6
  </button>
  <button
    className="Operations"
    name="-"
    onClick={[Function]}
    type="button"
  >
    -
  </button>
  <button
    className=""
    name="1"
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    className=""
    name="2"
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    className=""
    name="3"
    onClick={[Function]}
    type="button"
  >
    3
  </button>
  <button
    className="Operations"
    name="+"
    onClick={[Function]}
    type="button"
  >
    +
  </button>
  <button
    className=""
    name="0"
    onClick={[Function]}
    type="button"
  >
    0
  </button>
  <button
    className=""
    name="."
    onClick={[Function]}
    type="button"
  >
    .
  </button>
  <button
    className=""
    name="="
    onClick={[Function]}
    type="button"
  >
    =
  </button>
</div>
`;
